### ISSUE TITLE
Potential Integration with Pyrender PBR

### ISSUE THREAD

**Issue by mmatl** [mikedh/trimesh#301]
Hi @mikedh!

I've created a new repo for my changes to `meshrender`. For now, I called it [pyrender](https://github.com/mmatl/pyrender) -- open to suggestions for a more catchy name :) 

Hoping to get a bit of feedback on what works and what you'd like to see change for it to integrate nicely with trimesh. Here are some points that I know will need work:

- For now, it seems to work well on Ubuntu! Performance can take a hit when shadowmapping for many lights -- I can fix this with a deferred rendering pipeline, but that would have to come at a later date when I have some time to write one.
- The repo uses GLSL 3+ shaders, so you need to be able to get a core profile in Pyglet to use the viewer. This works without difficulty on Linux, but Pyglet's code for MacOS is broken. I've submitted PRs to try to fix this, but they're quite slow to respond and release new versions. I forked Pyglet and can point the setup.py to use my fork, but it's not the prettiest solution.
- Shadowmapping still doesn't work nicely on MacOS (something to do with retina displays, will need to debug).
- I still need to test things on Windows.

Best,
Matt

**Comment by mikedh** [comment]
Hey Matt, sounds great! I like the new name haha, and I see you registered it on pypi 👍.  How does this sound:
## Integration Plan
- `trimesh` will add `pyrender` to `pip install trimesh[all]` deps
- `trimesh.Scene.show` will add another viewer option, `pyrender`. Defaults to it if the scene only contains `Trimesh` objects (unless you want to support `PointCloud` and `Path` objects)
- `trimesh.save_image` will add a viewer option, `pyrender`


### Platform Notes
#### Windows
- all the deps install cleanly, except for `pyopengl_accelerate` which wants to compile stuff (and that's kind of a mud wrestling match in windows). 
  - 🙏 they have windows wheels in [pre- release](https://pypi.org/project/PyOpenGL-accelerate/3.1.3b1/#files) which I'll try later, and hopefully get released soon.
- I tried it on windows without `pyopengl_accelerate`, and wasn't able to get an OpenGL 3+ context due to something WGL_ARB. Possibly solved by the missing dep.

#### Mac
- re: pyglet, have you tried PR'ing into their main [bitbucket repo](https://bitbucket.org/pyglet/pyglet/overview)? I submitted a minor fix a few years ago and they were pretty responsive.

**Comment by mikedh** [comment]
Sounds good! Yeah "slow rolling" sounds good, I'd still like to make some improvements in the texture importer (specifically on freaking OBJ files), which might change the `SimpleMaterial` structure slightly. Maybe we should shoot for doing a release with both in late Febuary or early March?

I'll ping the maintainer of `pyopengl_accelerate` to see if he's planning on doing a release of the pre-release wheels. Your list looked good, here's the stuff I hope to tackle before then:

#### MDH
- Have Shapenet OBJ models load
- Have GLTF export texture
- Figure out something so you don't have to do `process=False` for textured models to not look like garbage. Maybe save the original unmerged vertices to cache as `smoothed` or something

#### Upstream
- `pyopengl_accelerate` windows wheels
- `pyglet` merge macos fix

**Comment by mmatl** [comment]
PR was merged into Pyglet's default branch. Will wait on the next PyPi release.

**Comment by mmatl** [comment]
Sorry, just getting around to looking at this again! I'm happy to add that function to `pyrender` -- looks awesome! Thanks for writing it out :)

**Comment by mmatl** [comment]
Added this to pyrender here: https://github.com/mmatl/pyrender/commit/ed4333ff2d29014e9353b727e0dd39e62e0a9f3d.

**Comment by jackd** [comment]
Since 3.0 the trimesh viewer is visualizing `obj`s very nicely. Unfortunately, I have two issues with it:

* lighting doesn't seem to be customizable? Or maybe I'm misunderstanding... but when I add custom lights I get indistinguishable results whether the intensity is 10.0 or 0.0.
* offscreen rendering is broken (I keep getting saved screenshots of my desktop).

`pyrender` seems to resolve these, but it looks like the work that went into better obj loading/more generalized textures doesn't carry across. I'm happy to try and hack something together with a bit of direction, but I'm really not a graphics guy...

Example below will download a shapenet category to `tmp/shapenet_data/SYNSET_ID.zip` - the hard-coded synset_id (for caps) is pretty small.

```python
import os
import pyrender
import trimesh
import wget
import zipfile
import numpy as np


class ZipSubdirResolver(trimesh.visual.resolvers.Resolver):
    def __init__(self, archive, subdir):
        assert(hasattr(archive, 'read'))
        self.archive = archive
        self.subdir = subdir

    def get(self, name):
        name = name.lstrip()
        if name.startswith('./'):
            name = name[2:]
        with self.archive.open(os.path.join(self.subdir, name)) as fp:
            return fp.read()


def compare(zf, synset_id, model_id):
    resolution = (512, 512)
    camera_transform = np.array(
        [[0.58991882, -0.3372407,   0.73366511,  0.86073076],
         [0.04221561,  0.92024442,  0.38906047,  0.45644301],
         [-0.80635825, -0.19854197,  0.55710632,  0.65359322],
         [0.0,         0.0,         0.0,         1.0]]
    )
    fov = 45
    light_kwargs = dict(
        color=np.ones(3),
        intensity=10.0,
        innerConeAngle=np.pi/16.0)

    print('Visualizing %s / %s' % (synset_id, model_id))
    subdir = os.path.join(synset_id, model_id)
    obj_path = os.path.join(subdir, 'model.obj')
    print(subdir, obj_path)
    resolver = ZipSubdirResolver(zf, subdir)

    with zf.open(obj_path) as fp:
        mesh_or_scene = trimesh.load(fp, file_type='obj', resolver=resolver)
        if isinstance(mesh_or_scene, trimesh.Trimesh):
            meshes = [mesh_or_scene]
        else:
            assert(isinstance(mesh_or_scene, trimesh.Scene))
            meshes = mesh_or_scene.geometry.values()
    camera = trimesh.scene.cameras.Camera(resolution=resolution, fov=(fov,)*2)
    light = trimesh.scene.lighting.SpotLight(**light_kwargs)
    tm_scene = trimesh.scene.Scene(
        camera=camera,
        camera_transform=camera_transform,
        geometry=meshes,
        lights=[light])
    tm_scene.graph[light.name] = camera_transform

    pr_scene = pyrender.Scene.from_trimesh_scene(tm_scene)

    if isinstance(tm_scene, trimesh.Trimesh):
        pr_scene.add(pyrender.Mesh.from_trimesh(tm_scene))
        tm_scene = tm_scene.scene()
    else:
        for mesh in tm_scene.geometry.values():
            pr_scene.add(pyrender.Mesh.from_trimesh(mesh))

    dist = np.linalg.norm(camera_transform[:3, 3])

    camera = pyrender.PerspectiveCamera(
        yfov=np.pi / 180 * fov,
        znear=dist-0.5, zfar=dist+0.5
    )

    print('trimesh')
    tm_scene.show()

    camera_transform = np.array(tm_scene.camera_transform)
    pr_scene.add(camera, pose=camera_transform)
    light = pyrender.SpotLight(**light_kwargs)
    pr_scene.add(light, pose=camera_transform)
    print('pyrender')
    pyrender.Viewer(pr_scene, viewport_size=resolution)


DL_URL = (
    'http://shapenet.cs.stanford.edu/shapenet/obj-zip/ShapeNetCore.v1/'
    '{synset_id}.zip')

# synset_id = '02773838'  # suitcases
synset_id = '02954340'    # cap
dl_url = DL_URL.format(synset_id=synset_id)
dl_dir = os.path.join('/tmp', 'shapenet_data')
if not os.path.isdir(dl_dir):
    os.makedirs(dl_dir)
path = os.path.join(dl_dir, os.path.split(dl_url)[1])

if not os.path.isfile(path):
    wget.download(DL_URL.format(synset_id=synset_id), out=path)


with zipfile.ZipFile(path, 'r') as zf:
    model_ids = [
        n.split('/')[1] for n in zf.namelist() if n.endswith('model.obj')]
    for model_id in model_ids:
        compare(zf, synset_id, model_id)
```

**Comment by jjmontesl** [comment]
> trimesh.Scene.show will add another viewer option, pyrender.

Was this finally implemented?
